minetest/minetest
# [manual merge] VBO-related optimizations and improvements

~~:warning: requires https://github.com/minetest/irrlicht/pull/292~~

**goal**:
* use more VBOs (not constantly uploading data to the GPU is supposed to be good for performance)
* move cloud rendering away from immediate mode and use a shader for further optimization
* some code cleanup

**applicability**: dunno, but I had fun and learned stuff while working on the cloud code.

**note for second commit**:
We would need to wrap every call to `setHardwareMappingHint` (try ctrl+f) into an if that checks the setting and also cache that etc. All while there is essentially no reason to disable VBOs.
(Irrlicht won't use them if they are not supported for some reason.)

## How to merge

Keep all commits that start with upper-case letters
Squash others into one and name it `VBO-related optimizations and improvements (#14395)`

## To do

This PR is Ready for Review.

## How to test

1. check and make sure dynamic shadows still work
2. open main menu, observe clouds
3. use test script and enjoy rapidly changing cloud cover (join a mgv6 world or MTG will get in the way)
```lua
local passed = 0
core.register_globalstep(function(dtime)
	passed = passed + dtime
	if passed < 0.4 then
		return
	end
	passed = passed - 0.4

	local t = core.get_us_time() / (1000*1000)
	for _, player in ipairs(minetest.get_connected_players()) do
		player:set_clouds({
			density = (1 + math.sin(t)) / 2
		})
	end
end)
```
4. watch cloud colors at dusk (`/time 19:00`)
5. check if anything is actually faster than before?
